File: tests/data/http/httpStatus.test.ts
import { isSuccess, isRetryable } from "@data/http/httpStatus";

/**
 * C-06 — **HTTP 206 est un succès.** `fetch` ne lève pas sur un 206, donc un
 * `if (res.status === 200)` passe la revue de code et casse à la première
 * pagination.
 *
 * Constaté trois fois en production cette session, sur trois endpoints :
 *   referentiel/stations?size=10000  → 200
 *   observations_tr?size=3           → 206
 *   ecoulement/observations?size=1000 → 206
 * Le même endpoint rend 200 ou 206 selon `size`. On ne peut donc pas décider
 * par endpoint : il faut accepter les deux, à un seul endroit.
 */
describe("statuts de succès Hub'Eau (C-06)", () => {
  it("accepte 200", () => {
    expect(isSuccess(200)).toBe(true);
  });

  it("accepte 206 — une réponse partielle est un succès paginé", () => {
    expect(isSuccess(206)).toBe(true);
  });

  it("refuse les statuts que l'on a réellement rencontrés en erreur", () => {
    expect(isSuccess(400)).toBe(false); // ValidatePageSize au-delà de size=10000
    expect(isSuccess(403)).toBe(false); // l'API v1 hydrométrie, arrêtée (C-01)
    expect(isSuccess(409)).toBe(false); // VigiEau ?commune= (C-14)
    expect(isSuccess(500)).toBe(false);
  });

  it("refuse les autres 2xx plutôt que de les supposer bons", () => {
    // 204 n'a pas de corps : le traiter en succès ferait échouer le parsing
    // plus loin, avec un message sans rapport.
    expect(isSuccess(204)).toBe(false);
    expect(isSuccess(202)).toBe(false);
  });
});

describe("statuts qui méritent une nouvelle tentative (C-15)", () => {
  it("réessaie sur 429 et sur les 5xx", () => {
    // Hub'Eau n'annonce aucun quota chiffré : le 429 est possible et transitoire.
    expect(isRetryable(429)).toBe(true);
    expect(isRetryable(500)).toBe(true);
    expect(isRetryable(503)).toBe(true);
  });

  it("ne réessaie jamais sur une erreur de requête", () => {
    // 400 vient de NOTRE requête : réessayer ne ferait que marteler un service
    // public gratuit sans aucune chance de succès.
    expect(isRetryable(400)).toBe(false);
    expect(isRetryable(403)).toBe(false);
    expect(isRetryable(409)).toBe(false);
    expect(isRetryable(404)).toBe(false);
  });

  it("ne réessaie pas sur un succès", () => {
    expect(isRetryable(200)).toBe(false);
    expect(isRetryable(206)).toBe(false);
  });
});

File: test/data/http/http_status_test.dart
// Verrouille ce qu'est un succès HTTP à un seul endroit (C-06). Constaté le
// 2026-09-13 sur le même endpoint hydrométrie : size=2 renvoie 206, une
// réponse vide renvoie 200 — un client qui n'accepte que 200 casse dès la
// première pagination. 204/304 ne sont pas des échecs mais n'ont pas de corps
// JSON : les compter en succès ferait échouer la désérialisation plus loin,
// pas la lecture du statut.
import 'package:flutter_test/flutter_test.dart';
import 'package:martinpecheur/data/http/http_status.dart';

void main() {
  group('isSuccess (C-06)', () {
    test('200 et 206 sont des succès', () {
      expect(isSuccess(200), isTrue);
      expect(isSuccess(206), isTrue);
    });

    test("204 et 304 n'apportent pas de corps JSON, ce ne sont pas des "
        'succès traités ici', () {
      expect(isSuccess(204), isFalse);
      expect(isSuccess(304), isFalse);
    });

    test('les 4xx et 5xx ne sont pas des succès', () {
      for (final int code in <int>[400, 403, 404, 409, 429, 500, 503]) {
        expect(isSuccess(code), isFalse, reason: '$code ne doit pas réussir');
      }
    });
  });

  group('isRetryable (C-12, C-01, C-14)', () {
    test('429 est rejouable (C-12) — aucun quota chiffré', () {
      expect(isRetryable(429), isTrue);
    });

    test('les 5xx sont rejouables', () {
      for (final int code in <int>[500, 502, 503, 599]) {
        expect(isRetryable(code), isTrue, reason: '$code doit être rejouable');
      }
    });

    test('les 4xx viennent de notre requête, aucun n\'est rejouable '
        '(403 = C-01, 409 = C-14)', () {
      for (final int code in <int>[400, 401, 403, 404, 409, 422]) {
        expect(
          isRetryable(code),
          isFalse,
          reason: '$code ne doit pas être rejouable',
        );
      }
    });

    test('200 et 206 ne sont pas rejouables, ce sont déjà des succès', () {
      expect(isRetryable(200), isFalse);
      expect(isRetryable(206), isFalse);
    });
  });
}
